ironfish/src/memPool/recentlyEvictedCache.ts

```diff
@@ -0,0 +1,218 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { Logger } from '../logger'
+import { MetricsMonitor } from '../metrics'
+import { TransactionHash } from '../primitives/transaction'
+import { PriorityQueue } from './priorityQueue'
+
+interface EvictionQueueEntry {
+  hash: TransactionHash
+  feeRate: bigint
+}
+
+interface ExpireAtQueueEntry {
+  hash: TransactionHash
+  expireAtSequence: number
+}
+
+/**
+ * A cache to track transactions that have recently been evicted from the mempool.
+ *
+ * This is primarily a mechanism to prevent transactions from being re-added to the mempool
+ * within a reasonable duration if for some reason they are evicted (i.e. the mempool is full
+ * and the transaction is underpriced).
+ *
+ * When full, transactions are evicted in order of `feeRate` descending to make room for the
+ * new transaction. This order is chosen because these transactions most likely to be successfully
+ * re-introduced into the mempool
+ */
+export class RecentlyEvictedCache {
+  private readonly logger: Logger
+
+  private readonly metrics: MetricsMonitor
+
+  /**
+   * The maximum size of the cache in number of transactions.
+   */
+  public readonly maxSize: number
+
+  /**
+   * A priority queue of transaction hashes ordered by `feeRate` descending.
+   * This is the eviction order for the cache when full.
+   */
+  private readonly evictionQueue: PriorityQueue<EvictionQueueEntry>
+
+  /**
+   * A priority queue of transaction hashes ordered by their expiration sequence.
+   */
+  // TODO(holahula): this can just be a normal queue because if y is inserted after x, y cannot expire before x
+  private readonly expireAtQueue: PriorityQueue<ExpireAtQueueEntry>
+
+  /**
+   * Creates a new RecentlyEvictedCache.
+   * Transactions that are evicted from the mempool should be added here.
+   *
+   * @constructor
+   * @param options.capacity the maximum number of hashes to store in the cache
+   */
+  constructor(options: { logger: Logger; metrics: MetricsMonitor; maxSize: number }) {
+    this.maxSize = options.maxSize
+    this.metrics = options.metrics
+    this.logger = options.logger.withTag('RecentlyEvictedCache')
+
+    this.metrics.memPool_RecentlyEvictedCache_MaxSize.value = this.maxSize
+
+    this.evictionQueue = new PriorityQueue<EvictionQueueEntry>(
+      (t1, t2) => t1.feeRate > t2.feeRate,
```

Comment:
should this use `mempoolEntryComparator`?

Reply:
yea good call